### DISCUSSION TITLE
Canceled (debouncer) search network-request with non-disposable provider shows the cached empty data

### DISCUSSION THREAD

**Discussion by iamsahilsonawane** [rrousselGit/riverpod#1695] [Q&A] [answered]
I was having an issue with the pagination and search, which got fixed, more info on https://github.com/rrousselGit/riverpod/discussions/1687

This is a relatively following issue.

## Problem Statement
I am attaching my current code for working pagination and debounced search below. Basically, it's canceling the network requests made between debouncer duration. 

Now when I cancel the request, I send a list with empty items for all the calls which are canceled which causes them to cache.

## Example
An example of a problem with this is when I type in 4 letter search, it debounces and cancels the network requests until the 3rd letter and thus returns empty data. 

For the fourth letter, it fetches from the API, but now, it shows the data if the 4 letters are present but if I go back to 3 or 2 letters it still shows empty results for an obvious reason i.e. it's cached.

Any workaround to this?

## Current solution

```dart
final paginatedVendorListProvider =
    FutureProvider.family<PaginatedResponse<Vendor>, SearchPaginationArgs>(
        (ref, pageData) async {
  ref.watch(refreshVendorListProviders);

  final adminRepo = ref.watch(adminRepositoryProvider);

  final cancelToken = ref.cancelToken();
  var debouncing = true;
  ref.onCancel(() {
    if (debouncing) cancelToken.cancel();
  });
  await ref.debounce(const Duration(milliseconds: 250));
  debouncing = false;

  if (cancelToken.isCancelled) {
    return const PaginatedResponse<Vendor>(
      data: Data(additionalData: [], list: [], paginationData: null),
    );
  }

  log("Fetching from API");

  return await adminRepo.fetchAllVendors(
      pageNumber: pageData.page!,
      search: pageData.search!,
      cancelToken: cancelToken);
});
```

## Demo

https://user-images.githubusercontent.com/47249618/192148351-d3fde15b-5f05-42bf-a612-0c738be405f6.mp4

**Comment by TimWhiting** [comment, marked as answer]
I believe there is a `ref.onResume` that you can use along with `ref.invalidateSelf` to cause it to recompute the state. Add that hook when you end up canceling.

**Reply by iamsahilsonawane** [reply]
Thanks for the answer. Now that I think about this way, just calling `ref.invalidateSelf` on the cancellation works fine as well. Now it works perfectly fine!

**Reply by iamsahilsonawane** [reply]
For others, I just added `ref.invalidateSelf` which invalidates the dummy states which are used to cancel out network-requests. 
```dart
...
  if (cancelToken.isCancelled) {
    ref.invalidateSelf(); // invalidate dummy providers.
    return const PaginatedResponse<Vendor>(
      data: Data(additionalData: [], list: [], paginationData: null),
    );
  }
...
```
